velox/docs/develop/debugging/tracing.rst

```diff
@@ -0,0 +1,146 @@
+
+**Replaying Phase**
+
+- Read and deserialize the recorded query plan in tracing phase, extract the target plan
+  node, and assemble a plan fragment with a customized source and sink nodes. The source
+  node reads the input from the serialized operator inputs on storage and sink operator
+  prints or logs out the execution stats.
+- Build a query/task with the assembled plan fragment in step 1.
+- Apply the recorded query configuration and connector properties to replay the query/task
+  with the same input and configuration setup as in production.
+
+**NOTE**: the presto serialization might lose the input vector encoding such as lazy vector
+and nested dictionary encoding etc which affects the operator’s execution. So it might not
+always be the exactly the same as in production.
+
+.. image:: ../images/trace-arch.png
+    :width: 600
+    :align: center
+
+Tracing Framework
+-----------------
+
+The tracing framework consists of three components:
+
+1. **Query Trace Writer**: metadata writer and the data writer.
+2. **Query Trace Reader**: metadata reader and the data reader.
+3. **Query Trace Tool**: display query summaries or replay the
+   execution of the target operator.
+
+Query Trace Writer
+^^^^^^^^^^^^^^^^^^
+
+**QueryMetadataWriter** records the query metadata during task creation,
+serializes, and writes them into a file in JSON format. There are two kinds
+of metadata:
+
+- Query configurations and connector properties are specified by the user per query.
+  They can be serialized as JSON map objects (key-value pairs).
+- Plan fragment of the task (also known as a plan node tree). It can be serialized
+  as a JSON object, which is already supported in Velox.
+
+**QueryDataWriter** records the input vectors from the target operator, which are
+serialized and written as a binary file.
+
+Query Trace Reader
+^^^^^^^^^^^^^^^^^^
+
+**QueryMetadataReader** can load the query metadata JSON file, and extract the query
+configurations, connector properties, and the plan fragment.
+
+**QueryDataReader** can read and deserialize the input vectors of the target operator.
+It is used as the utility to replay the input data as a source operator in the target
+operator replay.
+
+**NOTE**: `QueryDataWriter` serializes and flushes the input vectors in batches,
+allowing it to replay the input process using the same sequence of batches.
+
+Query Trace Util
+^^^^^^^^^^^^^^^^
+
+- Create tracing directories.
+- Get query summaries.
+- Provide utilities to extract the target plan node, and assemble a plan fragment with
+  customized trace source and sink node with it.
+
+Query Trace Tool
+----------------
+
+The query trace tool leverages the trace reader to display query summaries and replay the
+execution of specific operators (TBD).
+
+Tracing tools Usage
+^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: c++
+
+  query_trace_tool --root $root_dir --summary --pretty
```

Comment:
Is replay a separate tool?

I still think that showing plan as JSON is not very useful. Can we show it in a more user-friendly format?

Reply:
>Is replay a separate tool?

No, I forget to update the name, it should be `query_replayer`.

>I still think that showing plan as JSON is not very useful. Can we show it in a more user-friendly format?

The plan is recorded during the task creation so no stats information in it. It is a bonus that allows users to get a preliminary understanding of the tracked data before replaying the query.

Reply:
Should we remove the plan JSON showing, and only list the traced task ids? @mbasmanova @xiaoxmeng

Reply:
It would be nice to allow for showing the plan, just not using JSON. Why can't we display it similar to printPlanWithStats, just without the stats?

Reply:
Got it, we need to use a human-friendly way to show the plan instead of using JSON. It should be similar to the way used in the `printPlanWithStats` although without the stats. Do I understand correctly? @mbasmanova

Reply:
Yup.

Reply:
Fixed by using `queryPlan->toString(true, true)`. Now the plan showing is human-friendly as the follows :)
```C++
-- HashJoin[5][INNER c0=u0, filter: lt(ROW["c0"],135)] -> c0:BIGINT, c1:SMALLINT, c2:TINYINT
  -- Project[1][expressions: (c0:BIGINT, ROW["c0"]), (c1:SMALLINT, ROW["c1"]), (c2:TINYINT, ROW["c2"])] -> c0:BIGINT, c1:SMALLINT, c2:TINYINT
    -- Values[0][1 rows in 1 vectors] -> c0:BIGINT, c1:SMALLINT, c2:TINYINT, c3:VARCHAR
  -- Project[4][expressions: (u0:BIGINT, ROW["c0"]), (u1:SMALLINT, ROW["c1"]), (u2:TINYINT, ROW["a0"])] -> u0:BIGINT, u1:SMALLINT, u2:TINYINT
    -- Aggregation[3][SINGLE [c0, c1] a0 := min(ROW["c2"])] -> c0:BIGINT, c1:SMALLINT, a0:TINYINT
      -- Values[2][1 rows in 1 vectors] -> c0:BIGINT, c1:SMALLINT, c2:TINYINT, c3:VARCHAR
```